### PR TITLE
Answer storage mutability-related questions with requests

### PR DESCRIPTION
**Pull request by slavapestov**
Builds on https://github.com/apple/swift/pull/25588, replacing a bunch of logic in `validateDecl()` with requests that are run automatically when these accessors are called on FuncDecl or AbstractStorageDecl:

- FuncDecl::getSelfAccessKind()
- AbstractStorageDecl::isGetterMutating()
- AbstractStorageDecl::isSetterMutating()
- AbstractStorageDecl::getOpaqueReadOwnership()

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test compiler performance

**Comment by slavapestov** [comment]
@swift-ci Please test Linux

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -4455,15 +4452,20 @@ class AbstractStorageDecl : public ValueDecl {
     Bits.AbstractStorageDecl.SupportsMutation = implInfo.supportsMutation();
   }
 
+  struct {
+    unsigned IsGetterMutatingComputed : 1;
+    unsigned IsGetterMutating : 1;
+    unsigned IsSetterMutatingComputed : 1;
+    unsigned IsSetterMutating : 1;
+  } LazySemanticInfo = { };
```
Comment (xymus) [inline include/swift/AST/Decl.h]: What makes you prefer to have these fields in `LazySemanticInfo` than in the `Bits`?

**Review by xymus — APPROVED** [review]

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (slavapestov) [reply include/swift/AST/Decl.h]: It means I can put the 'friend class FooRequest;' inside AbstractStorageDecl and not ValueDecl. It doesn't really matter though, and putting it in Bits would save memory.

@DougGregor any thoughts?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Pull request merged**